src/composite-project.ts

```diff
@@ -0,0 +1,103 @@
+import * as path from 'path';
+import { Component } from './component';
+import { Project } from './project';
+
+/**
+ * Options for `CompositeProject`.
+ */
+export interface CompositeProjectOptions {
+  /**
+   * Declaratively define sub-projects by their sub paths.
+   * @example ```ts
+   * [
+   *   {
+   *     path: path.join('packages', 'foo'),
+   *     project: new NodeProject({ name: 'foo' }),
+   *   },
+   *   ...
+   * ]
+   * ```
+   */
+  projects?: CompositeProjectDeclarativeSubProject[];
+}
+
+/**
+ * Declares a sub-project of the composite project.
+ */
+export interface CompositeProjectDeclarativeSubProject {
+  /**
+   * Project subpath.
+   */
+  readonly path: string;
+
+  /**
+   * Projen project to synthesize in `path`.
+   */
+  readonly project: Project;
+}
+
+/**
+ * Creates a composite project.
+ */
+export class CompositeProject extends Project {
+  // Tracks which sub paths have already been added.
+  private readonly projects: Record<string, ProjectComponent>;
+
+  constructor(options?: CompositeProjectOptions) {
+    super();
+
+    this.projects = {};
+
+    // Add declaratively defined subprojects.
+    for (const subProject of options?.projects ?? []) {
+      this.addProject(subProject.path, subProject.project);
+    }
+  }
+
+  /**
+   * Adds a project as a sub-project at a sub path.
+   * @param subPath
+   * @param project
+   */
+  addProject(subPath: string, project: Project) {
+    if (this.projects[subPath]) {
+      throw new Error(`Cannot add project as the sub path ${subPath} is already in use`);
+    }
+
+    this.projects[subPath] = new ProjectComponent(this, {
+      path: subPath,
+      project: project,
+    });
+  }
+}
+
+/**
+ * Options for `ProjectComponent`
+ * @experimental
+ */
+export interface ProjectComponentOptions {
```

Comment:
Can we leave this private for now? Feels like it could be trouble and I am not sure I fully understand the use case here.

```suggestion
interface ProjectComponentOptions {
```